File: homework2observer/src/main/java/by/itacademy/homework2observer/MainActivity.java
package by.itacademy.homework2observer;

import android.content.Intent;
import android.os.Bundle;
import android.util.Log;
import android.view.View;

import androidx.appcompat.app.AppCompatActivity;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity implements IObserver {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        Log.d("HM2", "MainActivity create");
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        Publisher.getInstance().addSubscriber(this);
        findViewById(R.id.startSecondActivity).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Publisher.getInstance().setData(RandomSetNumbers.createArray(100));
                startActivity(new Intent(MainActivity.this, SecondActivity.class));
            }
        });
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        Publisher.getInstance().removeSubscriber(this);
        Log.d("HM2", "MainActivity ONDESTROY");
    }

    @Override
    public void notifyDataChanged(String data, ArrayList<Integer> arrayList) {
        if (data != null) {
            Log.d("HM2", "MainActivity notifyDataChanged " + data);
        }
    }
}

File: homework9mvvm/src/main/java/by/itacademy/homework9mvvm/MainActivity.kt
package by.itacademy.homework9mvvm

import android.content.Intent
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import by.itacademy.homework9mvvm.databinding.ActivityMainBinding
import by.itacademy.homework9mvvm.model.HourlyWeatherModel
import by.itacademy.homework9mvvm.model.WeatherModel
import by.itacademy.homework9mvvm.presentation.HourlyWeatherAdapter
import by.itacademy.homework9mvvm.presentation.MainActivityViewModel


class MainActivity : AppCompatActivity() {
    private lateinit var binding: ActivityMainBinding
    private val weatherAdapter by lazy { HourlyWeatherAdapter() }

    private lateinit var mainActivityViewModel: MainActivityViewModel
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        initViewModel()
        with(binding) {
            recycler.apply {
                layoutManager = LinearLayoutManager(this@MainActivity)
                adapter = weatherAdapter
            }
            settingsButton.setOnClickListener { startActivity(Intent(this@MainActivity, SettingsActivity::class.java)) }
            cityButton.setOnClickListener { startActivity(Intent(this@MainActivity, CityActivity::class.java)) }
        }
    }

    override fun onStart() {
        super.onStart()
        mainActivityViewModel.getMainWeatherFromApi()
    }

    private fun initViewModel() {
        mainActivityViewModel = ViewModelProvider.AndroidViewModelFactory.getInstance(this.application).create(MainActivityViewModel::class.java)
        mainActivityViewModel.weatherLiveData.observe(this, Observer<WeatherModel> { weatherModel ->
            with(binding) {
                city.text = weatherModel.cityName
                degree.text = weatherModel.temp
                description.text = weatherModel.description
            }
        })
        mainActivityViewModel.weatherAdapterLiveData
                .observe(this, Observer<List<HourlyWeatherModel>> { hourlyweather -> weatherAdapter.update(hourlyweather) })
    }
}